### PR TITLE
adding firebase, firestore and auth refactoring

### PR DESCRIPTION
**Pull request by the-muppet**
we are all draft pr

### CODE REVIEW

File: firebase.go [hunk 1]
```diff
@@ -0,0 +1,147 @@
...
+
+	// Add user preferences to the Firestore document
+	_, err = docRef.Update(ctx, []firestore.Update{
+		{
+			Path:  "preference",
+			Value: "Preference value",
+		},
+	})
+	if err != nil {
+		log.Fatalf("docRef.Update: %v", err)
```
Comment (kodawah) [inline firebase.go]: fatalf will shut down the whole system on error, like the entire website, this needs to handle failures more gracefully

File: login.go [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+	}
+
+	token, err := client.VerifyIDToken(context.Background(), idToken)
+	if err != nil {
+		http.Error(w, "Unauthorized", http.StatusUnauthorized)
+		return
+	}
+
+	response := LoginResponse{
+		Message: fmt.Sprintf("Hello, %s! This is a protected route.", token.UID),
```
Comment (kodawah) [inline login.go]: can just return a 404 not found message, you typically dont want to reveal a route exists in the first place

File: main.go [hunk 3]
```diff
@@ -683,7 +687,7 @@ func main() {
 
 	http.Handle("/sets", enforceSigning(http.HandlerFunc(Search)))
 	http.Handle("/sealed", enforceSigning(http.HandlerFunc(Search)))
-
+	http.
```
Comment (kodawah) [inline main.go]: http?

File: main.go [hunk 4]
```diff
@@ -189,6 +189,10 @@ type NavElem struct {
 	CanPOST bool
 }
 
+type LoginResponse struct {
```
Comment (kodawah) [inline main.go]: you can move this type where it belongs (login.go) doesnt need to be in the same file as where it's used

File: main.go (hunk 3 above)
Comment (the-muppet) [reply main.go]: this was where i was going to add the handler but then ?? elmo happened

File: firebase.go (hunk 1 above)
Comment (the-muppet) [reply firebase.go]: un-fatalf'd

File: login.go (hunk 2 above)
Comment (the-muppet) [reply login.go]: wink

**Comment by the-muppet** [comment]
Tag you're it

**Pull request closed without merging**